src/polydatum/dal.py

```diff
@@ -1,20 +1,236 @@
+
+    def __call__(self, *args, **kwargs):
+        return self._handler(self.path, *args, **kwargs)
+
+
+class MethodMiddleware:
+    """
+    Base class used primarily for type checking and to provide
+    an interface for a method middleware.
+    """
+    def __call__(self, request: DalMethodRequest, handler: Callable):
+        pass
+
+
+def dal_resolver(ctx, dmr):
+    """
+    This function resolves a dal method call to its underlying
+    service
+    """
+    return resolve(DalMethodRequest(ctx=ctx, path=dmr.path, args=None, kwargs=None))
+
+
+def resolve(request: DalMethodRequest):
+    """
+    This function resolves a DalMethodRequest to a service method
+    """
+    service_or_method = request.ctx.dal._services  # noqa
+    paths = list(request.path[:])
+    location = []
+    while paths:
+        path_segment = paths.pop(0)
+        location.append(path_segment)
+        # Third time through, service_or_method might be `None`, but we want
+        # to continue walking the path. Everything after the first missing
+        # service/method will be `(location, None)`.
+        if service_or_method:
+
+            # This if condition is handling the case of the first loop here.
+            # we cannot use attribute access on the dal directly because of how
+            # attribute access is deferred with DalMethodRequest objects.
+            if isinstance(service_or_method, dict):
+
+                # If the code being resolved has typo'd a service name, this
+                # could be returning something that is not a service.
+                service_or_method = service_or_method.get(path_segment.name)
+            else:
+                # second time through (and beyond), service or method is a real
+                # service or method, and we do not need to use a special case for
+                # finding the first service.
+                try:
+                    service_or_method = getattr(service_or_method, path_segment.name)
+                    if not service_or_method and callable(service_or_method):
+                        raise DalMethodError(request, path=tuple(location))
+
+                except KeyError:
+                    raise DalMethodError(request, path=tuple(location))
+        else:
+            raise DalMethodError(request, path=tuple(location))
+    return service_or_method
+
+
+class DalMethodResolverMiddleware(MethodMiddleware):
+    """
+    A Method Middleware that resolves deferred dal attribute access to
+    underlying services/methods that has to be called
+    """
+
+    def __call__(self, request: DalMethodRequest, handler: Callable):
+        """
+        Args:
+             request: Input from caller
+             handler: Downstream middleware or actual DAL method handler
+        """
+        request.dal_method = resolve(request)
+        return handler(request)
+
+
+def default_handle_dal_method(request: DalMethodRequest):
```

Comment:
```suggestion
def handle_dal_method(request: DalMethodRequest):
```

I'm not sure "default" adds anything here.

Reply:
It makes it really explicit to me how it's being used.